plugins/libplugin-pay.c

```diff
@@ -768,16 +990,147 @@ retry_data_init(struct payment *p)
 	return rdata;
 }
 
+/* Determine whether retrying could possibly succeed. */
+static bool payment_can_retry(struct payment *p)
+{
+	struct payment_result *res = p->result;
+	u32 idx;
+	bool is_final;
+
+	if (p->result == NULL)
+		return false;
+
+	idx = res->erring_index != NULL ? *res->erring_index : 0;
+	is_final = (idx == tal_count(p->route));
+
+	/* Full matrix of failure code x is_final. Prefer to retry once too
+	 * often over eagerly failing. */
+	switch (res->failcode) {
+	case WIRE_EXPIRY_TOO_FAR:
+	case WIRE_INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS:
+	case WIRE_INVALID_ONION_PAYLOAD:
+	case WIRE_INVALID_ONION_VERSION:
+	case WIRE_INVALID_REALM:
+	case WIRE_MPP_TIMEOUT:
+	case WIRE_PERMANENT_NODE_FAILURE:
+	case WIRE_REQUIRED_NODE_FEATURE_MISSING:
+	case WIRE_TEMPORARY_NODE_FAILURE:
+	case WIRE_UNKNOWN_NEXT_PEER:
+		return !is_final;
+
+	case WIRE_AMOUNT_BELOW_MINIMUM:
+	case WIRE_CHANNEL_DISABLED:
+	case WIRE_EXPIRY_TOO_SOON:
+	case WIRE_FEE_INSUFFICIENT:
+	case WIRE_FINAL_INCORRECT_CLTV_EXPIRY:
+	case WIRE_FINAL_INCORRECT_HTLC_AMOUNT:
+	case WIRE_INCORRECT_CLTV_EXPIRY:
+	case WIRE_INVALID_ONION_HMAC:
+	case WIRE_INVALID_ONION_KEY:
+	case WIRE_PERMANENT_CHANNEL_FAILURE:
```

Comment:
i'm assuming 'retry the payment' means along a new route?

Reply:
Yes, retry the entire flow with a new sub-payment, and with updated channel hints which can then lead to the failing channel being excluded :+1:

Reply:
Added a comment about this to clarify.